qiskit/qpy/binary_io/schedules.py

```diff
@@ -0,0 +1,349 @@
+import json
+import struct
+import zlib
+
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+from qiskit.pulse import library, channels
+from qiskit.pulse.schedule import ScheduleBlock
+from qiskit.qpy import formats, common, type_keys
+from qiskit.qpy.binary_io import value
+from qiskit.utils import optionals as _optional
+
+
+def _read_channel(file_obj, version):
+    type_key = common.read_type_key(file_obj)
+    index = value.read_value(file_obj, version, {})
+
+    channel_cls = type_keys.ScheduleChannel.retrieve(type_key)
+
+    return channel_cls(index)
+
+
+def _read_waveform(file_obj, version):
+    header = formats.WAVEFORM._make(
+        struct.unpack(
+            formats.WAVEFORM_PACK,
+            file_obj.read(formats.WAVEFORM_SIZE),
+        )
+    )
+    samples_raw = file_obj.read(header.data_size)
+    samples = common.data_from_binary(samples_raw, np.load)
+    name = value.read_value(file_obj, version, {})
+
+    return library.Waveform(
+        samples=samples,
+        name=name,
+        epsilon=header.epsilon,
+        limit_amplitude=header.amp_limited,
+    )
+
+
+def _loads_symbolic_expr(expr_bytes):
+    from sympy import parse_expr
+
+    if expr_bytes == b"":
+        return None
+
+    expr_txt = zlib.decompress(expr_bytes).decode(common.ENCODE)
+    expr = parse_expr(expr_txt)
+
+    if _optional.HAS_SYMENGINE:
+        from symengine import sympify
+
+        return sympify(expr)
+    return expr
+
+
+def _read_symbolic_pulse(file_obj, version):
+    header = formats.SYMBOLIC_PULSE._make(
+        struct.unpack(
+            formats.SYMBOLIC_PULSE_PACK,
+            file_obj.read(formats.SYMBOLIC_PULSE_SIZE),
+        )
+    )
+    pulse_type = file_obj.read(header.type_size).decode(common.ENCODE)
+    envelope = _loads_symbolic_expr(file_obj.read(header.envelope_size))
+    constraints = _loads_symbolic_expr(file_obj.read(header.constraints_size))
+    valid_amp_conditions = _loads_symbolic_expr(file_obj.read(header.valid_amp_conditions_size))
+    parameters = common.read_mapping(
+        file_obj,
+        deserializer=value.loads_value,
+        version=version,
+        vectors={},
+    )
+    duration = value.read_value(file_obj, version, {})
+    name = value.read_value(file_obj, version, {})
+
+    # TODO remove this and merge subclasses into a single kind of SymbolicPulse
+    #  We need some refactoring of our codebase,
+    #  mainly removal of isinstance check and name access with self.__class__.__name__.
+    if pulse_type == "Gaussian":
+        pulse_cls = library.Gaussian
+    elif pulse_type == "GaussianSquare":
+        pulse_cls = library.GaussianSquare
+    elif pulse_type == "Drag":
+        pulse_cls = library.Drag
+    elif pulse_type == "Constant":
+        pulse_cls = library.Constant
+    else:
+        pulse_cls = library.SymbolicPulse
+
+    # Skip calling constructor to absorb signature mismatch in subclass.
+    instance = object.__new__(pulse_cls)
```

Comment:
Could you do `pulse_cls.__new__()` here for the same end result or do you have to use the object `__ new__` to avoid the signature mismatch?

Reply:
Yes, this is to avoid signature mismatch. Each symbolic pulse subclass doesn't take expressions (it's hardcoded in constructor). It's also possible to instantiate everything with SymbolicPulse and then override class but seems to me this is not clean.